docs: add lua 5.1 as part of the neovim documentation

dundargoc opened the pull request:
I think it'd be nice to have in core but I'm open to thoughts.

Snatched this from https://github.com/bfredl/luarefvim

Commit:
docs: add lua 5.1 as part of the neovim documentation

lewis6991 commented:
This keeps with the idiom that all documentation for Neovim is available offlin via the `:help` system.

If we do this, we should also add luv/libuv.

clason commented:
We might want to massage this somewhat, though -- as far as I can tell, this file uses additional syntax that requires(?) the syntax and plugin files distributed with it.

> If we do this, we should also add luv/libuv.

That is a good idea, but someone would have to create it first. Might as well turn it into a custom `vim.loop` documentation, then.

bfredl commented on runtime/doc/lua51refvim.txt:
```diff
@@ -0,0 +1,5617 @@
+                                   CONTENTS
+                                 ============
+
+           1   INTRODUCTION........................|lrv51-intro|
```
we probably change the prefix to something less keysmashy, maybe `luaref`, which also can be the filename. ("LuaRef" means something else in the C code base, but this term is not leaked to public docs, so there is no confusion)

bfredl commented on runtime/doc/help.txt:
```diff
@@ -185,6 +185,7 @@ Other ~
 |channel.txt|		Nvim asynchronous IO
 |dev_style.txt|		Nvim style guide
 |job_control.txt|	Spawn and control multiple processes
+|lua51refvim.txt|	Lua 5.1 reference manual
```
just Lua not "lua 5.1" in most places (except when in a direct quote like line 14).

bfredl commented:
Either needs to add these syntax classes or strip them out (likely some could folded into the standard help syntax groups)

```
" add three syntax classes: bold, emph (italic) and code -- similarly to html
"
" TODO: match luarefvim filename?
syn match helpIgnoreBold     "#[a-zA-Z0-9&()\`\'\"\-\+\*=\[\]\{\}\.,;: ]\+#" contains=helpMatchBold
syn match helpMatchBold      "[a-zA-Z0-9&()\`\'\"\-\+\*=\[\]\{\}\.,;: ]\+"   contained
syn match helpIgnoreEmph     "@[a-zA-Z0-9&()\`\'\"\-\+\*=\[\]\{\}\.,;: ]\+@" contains=helpMatchEmph
syn match helpMatchEmph      "[a-zA-Z0-9&()\`\'\"\-\+\*=\[\]\{\}\.,;: ]\+"   contained
" this match is the same as in CRefVim's help.vim (that is, uses $$):
" the idea is to keep some degree of portability.
" TODO: convert code blocks to standard < > notation
syn match helpIgnoreCode     "\$[a-zA-Z0-9@\\\*/\._=()\-+%<>&\^|!~\?:,\[\];{}#\`\'\" ]\+\$" contains=helpMatchCode
syn match helpMatchCode      "[a-zA-Z0-9@\\\*/\._=()\-+%<>&\^|!~\?:,\[\];{}#\`\'\" ]\+"   contained
```

lewis6991 commented:
> That is a good idea, but someone would have to create it first. Might as well turn it into a custom vim.loop documentation, then.

A plugin already exists for it. On mobile but will update later.

Commit:
docs: use luaref instead of lrv51 and drop the 51

dundargoc commented:
> A plugin already exists for it: [nanotee/luv-vimdocs](https://github.com/nanotee/luv-vimdocs)

hippity hoppity, your plugin is now my property :)

I've added it, thanks for the tip.

dundargoc commented:
I should mention that I haven't done any sort of proofreading of the documents yet, I wanted to get a feeling for the overall reception before putting in mountains of work :)

bfredl commented on runtime/doc/luarefvim.txt:
entire C.1 and C.2 are irrelevant when shipped in runtime, could just be deleted

Commit:
docs: remove D1 and D2 and adjust title alphabets

bfredl commented:
https://github.com/wsdjeg/luarefvim/blob/master/plugin/help.vim the last six  `hi link` here would also be needed, though ideally we will completely eliminate the special $ code blocks at least.

Commit:
refactor: delenda est $

Commit:
refactor: convert @

clason commented on runtime/doc/luarefvim.txt:
That's not correct anymore?

clason commented on runtime/doc/luarefvim.txt:
irrelevant?

clason commented on runtime/doc/help.txt:
```diff
@@ -185,6 +185,7 @@ Other ~
 |channel.txt|		Nvim asynchronous IO
 |dev_style.txt|		Nvim style guide
 |job_control.txt|	Spawn and control multiple processes
+|luarefvim.txt|		Lua reference manual
```
I'd use just `luaref.txt` (or `luareference.txt`, for disambiguation) -- the `vim` part is slightly misleading since the documentation is _not_ Vim-related (only the format is).

clason replied:
Maybe we should split off `api.txt`, `lua.txt`, and this (and the luv) reference manual to a separate section ("APIs")?

clason commented:
Do we want to (and can we) move the copyright to `LICENSE.txt`? There's a "external library" section there which could cover this. (@justinmk)

clason committed:
doc(lua): remove custom bold markup

clason committed:
doc(luaref): remove obsolete references

clason commented:
Did a bit of cleanup and removed all the weird bold markup.

clason committed:
docs(luaref): remove irrelevant standalone section

clason committed:
docs(lua): link to luaref on patterns

clason committed:
doc(lua): link package.{c,}path to luaref

clason committed:
docs(luaref): standard header, include upstream info

clason committed:
docs(luaref): rename to luarefman.txt

it's the default lua reference manual; the format is irrelevant in this
context (what else would it be if it's bundled?)

zeertzjq commented:
These are still many mentions of `luarefvim` instead of `luarefman`.

clason commented:
> These are still many mentions of `luarefvim` instead of `luarefman`.

Yes, I wanted to wait for feedback on the renaming before making (and having to undo) a lot of changes.

lewis6991 committed:
docs(luaref): format luarefman.txt to be consistent with lua.txt

- Right justified tags
- Matched formatting of lua.txt, specifically for functions.

lewis6991 commented:
I've done a significantly large reformat of `luarefman.txt` to be consistent with `lua.txt`. 

Let me know if anyone disagrees with anything I've done, or whether they would prefer the format be done more incrementally in multiple commits.

If we are happy with it, I can do the same for libuv.

clason commented:
No, better to do it right the first time. (The luv docs look fine as they are, though? @nanotee already did a great job.)

lewis6991 commented:
> No, better to do it right the first time. (The luv docs look fine as they are, though? @nanotee already did a great job.)

Haven't looked at them yet.

For reference, I used this lua to auto justify the tags:

```lua
local TW = 78

for i, l in ipairs(vim.api.nvim_buf_get_lines(0, 0, -1, false)) do
  if vim.endswith(l, '*') then
    local newline
    local _, nstars = l:gsub('*', '')

    -- Match padding + tags
    local suffix = l:match('%s*%*.*')
    local prefix = l:sub(1, -#suffix-1)

    if #l > TW then
      local offset = #l - TW - nstars

      -- Check we have enough padding to remove
      if vim.startswith(suffix, string.rep(' ', offset)) then
        newline = prefix..suffix:sub(offset+1)
      end
    elseif #l < TW then
      local offset = TW - #l + nstars
      newline = prefix..string.rep(' ', offset)..suffix
    end

    if newline then
      vim.api.nvim_buf_set_lines(0, i-1, i, false, {newline})
    end
  end
end
```

In case we want to tweak it further, or use this as a basis for other stuff.

clason commented:
I think you removed too much from the `LUAREFVIM DOC` heading? (File is too big to comment on it...)

lewis6991 committed:
fixup! bad automation

clason commented:
Now that we are already this close to native docs, we might as well fix the final nit for me: the colon in list items like

>  - `\a` : bell

lewis6991 committed:
fixup! where did lua_type go?

lewis6991 committed:
docs(luaref): tidy some lists

lewis6991 committed:
fixup! remove lua_type backticks

lewis6991 committed:
fixup! lua_type - properly close codeblock

lewis6991 committed:
docs(luaref): more list formatting

clason committed:
fixup! docs(luaref): rename to luarefman.txt

clason committed:
docs(luaref): remove inconsistent quoting in text

dundargoc commented:
Bit of a bikeshed: maybe just `luaref` instead of `luarefman`? Luarefman sounds like a bad super hero name (strong Macro Man vibes :) )

bfredl commented:
just luaref is fine. `LuaRef` is completely internal and thus not a name collision in public documentation.

Commit:
remove extra space

Commit:
rename luarefman to luaref

Commit:
Revert "remove extra space"

This reverts commit 17d2a7a11771fe11ad67cea17ae73368d06aec02.

dundargoc commented:
Took a quick look at it, looks awesome. Thanks for the work @clason and @lewis6991. I don't have any improvement ideas as you've already taken care of all of them :). I think it looks good enough to ship, but I'll let you decide.

Commit:
remove excessive code backticks

Commit:
remove section about lua sample program

clason committed:
fixup: formatting of inline tags

clason commented:
yay teamwork \o/
